Fix decoder for LGT-92 fw version 1.5/1.6 devices

gsauthof opened the pull request:
Especially, it fixes the issue where longitude/latitude are set to zero
if the MSB is zero.

It also fixes the decoding of the Move-Direction (MD) field, where e.g.
'Move' was erroneously decoded as 'Disable', before.

While at it I also added some warnings for some special conditions.

It's a complete rewrite of the previous decoder in order to make it
shorter, easier reviewable and arguably simpler to review.

fixes #331

<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->
This pull request fixes decoding errors in the latitude/longitude/move-detection fields, explicitly supports firmware 1.5/1.6 devices, simplifies the decoder and adds some warnings.

Fixes #331

#### Changes
<!-- What are the changes made in this pull request? -->

- fix latitude/longitude decoding logic with respect to values where the MSB is zero
- fix mapping of move-direction values
- skip optional fields if not present instead of setting them to `0` or `null`
- add warnings for special GPS values
- explicitly support firmware version 1.5/1.6 devices
- reduce decoder code size

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

Motivation: the decoding results of the existing decoder are buggy for my LGT-92 device.

The decoder now explicitly supports firmware version 1.5/1.6 devices and I stopped at support for 1.5 devices.
If there is a need I can also add support for firmware version 1.4 devices. But perhaps it's just and reasonable to just update the firmware on those devices.

I named the keys like in the existing decoder to keep them backwards compatible. But if there is a good reason I can also change them.

#### Release Notes
<!--
NOTE: This section is optional.

Any notes that we need to include in the Release Notes for the next release.
These notes are formatted as bullet points, written in past tense, and will be
combined with the labels of this Pull Request.

Always mention changes in API.
-->

- fixed: decoding of small longitude/latitude values and move-direction values
- fixed: `LON` and `ALARM_status` fields  are decoded as booleans instead of boolean-like strings
- fixed: optional fields such as altitude are decoded only if present in the payload

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/TheThingsNetwork/lorawan-devices?pullRequest=362) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/TheThingsNetwork/lorawan-devices?pullRequest=362) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/TheThingsNetwork/lorawan-devices?pullRequest=362) it.</sub>

Jaime-Trinidad commented:
@gsauthof thanks for your changes, please check the codec file and run `make validate` there are some error that need to be fix, please sign CLA

gsauthof commented:
The existing test case failed due to including bogus values for non-present optional fields, a wrongly decoded value,  representing booleans as strings and a trailing zero byte:

Status quo:

```
[0x01, 0x5A, 0x78, 0x54, 0x06, 0xCF, 0x33, 0xD8, 0x0D, 0x20, 0x60, 0x00, 0x4B, 0x01, 0x90, 0x00]
// i.e. 16 bytes
=>
{"data":{
"ALARM_status":"FALSE",
"altitude":0,
"BatV":3.36,
"FW":160,
"LON":"ON",
"latitude":22.70626,
"longitude":114.24252,
"MD":"Disable",
"Pitch":4,
"Roll":0.75,
"hdop":0
}}
```
vs. improved new version:

```
[0x01, 0x5A, 0x78, 0x54, 0x06, 0xCF, 0x33, 0xD8, 0x0D, 0x20, 0x60, 0x00, 0x4B, 0x01, 0x90]
// i.e. 15 bytes
=>
{
  data: {
    latitude: 22.70626,
    longitude: 114.24252,
    ALARM_status: false,
    BatV: 3.36,
    MD: 'Move',
    LON: true,
    FW: 160,
    Roll: 0.75,
    Pitch: 4
  }
}
```
Note that the 1.5 LGT92 manual specifies the full payload (including optional fields) length as 15 bytes - which also matches the included example.

I thus updated my pull request by fixing the reference data for this test case.

While at it I added a few additional test cases to increase the test coverage.

I also updated the decoder such that the `warnings` key is only returned if there are actually warnings - and I fixed the detection of a turned off GPS module.

---

The problem with the CLA is that it looks like it's written for lawyers and I'm not a lawyer.
While reading it my head started to hurt.
In contrast to open source licenses, where there are just a few standard licenses, each project that thinks it needs a CLA rolls their own.
I mean I don't even know if it was written by a competent lawyer, perhaps it has some accidental implications not even indented by its creators.
For example, some of its terms seem overly broad - e.g. does it even require that a work copyrighted by me is submitted (conveyed?!?) by me? Or is it possible that - after I've signed that CLA - that some random dude takes some arbitrary piece of code I've published in some github repository and conveys it by himself to a TTN repository such that me having signed the CLA would also grant the TTN project's maintainers/contributors/users/... an arbitrary irrevocable copyright license for such a contribution?

I have no problems contributing my changes under an ultra permissive license such as MIT-0 or Unlicense, which are both approved by the OSI.

Perhaps this would be an alternative?

---

I see that your CLA also contains some patent clauses - whereas the MIT-0/Unlicense don't address patents. However, perhaps we can agree that such trivial decoder code can't possibly be affected by any existing patent.

If we talk about more serious pieces of code, there are even some standard open source licenses that include patent clauses ...

Jaime-Trinidad commented:
> that some random dude takes some arbitrary piece of code I've published in some github repository and conveys it by himself to a TTN repository such that me having signed the CLA would also grant the TTN project's maintainers/contributors/users/... an arbitrary irrevocable copyright license for such a contribution?

Thanks for the fixes, as you know this is an open source repository, contributors as you can propose and make changes for a better work of devices, device makers or users that clone repository can work with the code you add or change, and we will review it, if it make sense (like this case) and after reviews we make changes, if you don't want to sign it we can't merge this PR. Let me know what you think.

Jaime-Trinidad commented:
Hi @gsauthof  any update here?

Jaime-Trinidad commented:
Hi @gsauthof please let me know about CLA

gsauthof commented:
No, I don't want to sign the CLA.
If there is no alternative way of contributing then it's like it is.

See also https://ben.balter.com/2018/01/02/why-you-probably-shouldnt-add-a-cla-to-your-open-source-project/ for a good summary of why a CLA is likely superfluous, how it scares away contributors and what better alternatives are available.
